Add in Micromobility

evansiroky opened the pull request:
I realize it is somewhat absurd to request a full PR review for this, but in the interest of moving things along, here are some bullet points of things that I'd like feedback on in terms of whether the approach seems correct.

One thing to keep in mind that this is all a bit rushed. I have copy-pasted and adapted a lot of the car rental code which itself is copy-pasted and adapted from the bike rental code. In a next phase of work I intend to merge bike, car and micromobility rental into a common set of classes.

## Having a new Mircomobility mode in the first place.

I can hear it now... maybe we could've just used the existing bicycling implementation to model eScooters! Close enough, right?! But, I'd like to present a few arguments as to why a new mode is needed:

- With a motor, I hypothesize that a user will behave differently than with a bicycle. With some more powerful eScooters they'll go as fast or faster than automobiles.
- Portland gives special treatment to eScooters in that they ban them from riding in parks while apparently regular bikes are ok
- I wanted to implement a dynamic way to calculate eScooter speed that took into account physics instead of the current ["black magic"](https://github.com/opentripplanner/OpenTripPlanner/issues/845#issuecomment-9158536) that [is used to calculate elevation-dependent bike speed](https://github.com/ibi-group/OpenTripPlanner/blob/1acab23b630942820c107998ce6761412925238f/src/main/java/org/opentripplanner/routing/util/ElevationUtils.java#L108-L236).

## New speed calculation for Micromobility mode

As noted in the last bullet point, a new physics-based caluclation of micromobility speed is implemented here: https://github.com/ibi-group/OpenTripPlanner/blob/3561e0e0dbf43c0cd99dac101a0b67a7817a5709/src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java#L857-L991

The elevation-dependent data for that equation is pre-computed in [ElevationUtils](https://github.com/ibi-group/OpenTripPlanner/blob/micromobility/src/main/java/org/opentripplanner/routing/util/ElevationUtils.java).

In order to be hopefully somewhat efficient with data storage and computation, the gradient data is stored as a list of distances encountered along a StreetWithElevationEdge in 1% bins of gradient. For example if a StreetWithElevationEdge had a total distance of 100 meters, but the first 50 meters were at a gradient of 1% and the second 50 meters were at 3%, then an array with two entries of the gradients and distances would be stored. See the [StreetWithElevationEdge.calculateSpeed](https://github.com/ibi-group/OpenTripPlanner/blob/3561e0e0dbf43c0cd99dac101a0b67a7817a5709/src/main/java/org/opentripplanner/routing/edgetype/StreetWithElevationEdge.java#L113-L159) method.

These speed calculations enable realistic fine-grained control over setting how fast a particular micromobility vehicle travels. New input parameters of `watts`, `weight`, `minimumMicromobilitySpeed` and `maximumMicromobilitySpeed` are used to control the speed of micromobility in graph searches.

## Adds ability to ban micromobility travel and rental parking via build-config.json

For now, all we got are these geojson files of where the City of Portland has banned eScooter travel and rental parking. Not sure where the ultimate place for these things should be... an updater or OSM, but for now they are loaded in via files specified in build-config.json. See [GraphBuilderParameters](https://github.com/ibi-group/OpenTripPlanner/blob/micromobility/src/main/java/org/opentripplanner/standalone/GraphBuilderParameters.java#L229-L230).

## GBFS refactor

The vehicle rental updater uses a completely new GBFS class. In the main OTP repo (https://github.com/opentripplanner/OpenTripPlanner/issues/2778), there is a desire to replace the bike rental GBFS updater with another one that makes more sense, so I went ahead and did that. As with the car rental updater, I also added in the ability to fetch a region wherein the company allows vehicle rental dropoffs. This is not yet part of the GBFS spec, so this may need to be refactored in the future.

evansiroky committed:
First pass at Micromobility speed calculations

evansiroky committed:
add more Micromobility speed tests and explain formula in Javadoc

evansiroky committed:
Add Micromobility icon

Image credit: https://www.flaticon.com/free-icon/kick_1254819

evansiroky committed:
Update UI to be able to plan with Micromobility mode

evansiroky committed:
Add ability to set min/max micromobility speeds as request params

evansiroky committed:
refactor micromobility speed calculation method to calculate just speed

evansiroky committed:
add test to make sure bicycling is not allowed on motorways by default

evansiroky committed:
use same bike walking settings for micromobility when walking with micromobility

evansiroky committed:
add micromobility permissions to ways with except tag

evansiroky committed:
set a low car speed for ways marked as having gravel or dirt surfaces

evansiroky committed:
first pass at calculating elevation data for micromobility

evansiroky committed:
avoid NPE on StreetWithElevationEdge when gradients weren't set

evansiroky committed:
First pass at adding scooter rentals

evansiroky committed:
Get some parts of vehicle rental working

evansiroky committed:
Get vehicle rentals working in departAt mode

evansiroky committed:
Get vehicle rental working in arriveBy mode

evansiroky committed:
Add VehicleRental API and add leg data

evansiroky committed:
Change default max micromobility speed to 5 m/s

evansiroky committed:
Speed up car rental region analysis

landonreed committed:
Merge remote-tracking branch 'origin/micromobility' into 2019-05-23

landonreed committed:
Merge branch 'pdx-fares' into 2019-05-23

Commit:
Refactor GBFS updater for vehicle rentals

Commit:
Delete old GBFS updater for vehicles

Commit:
improve logging by ensuring vertices and rental stations have unique names

Commit:
Get vehicle rentals working fully after some breakage

landonreed committed:
fix(graphql): add route#route_sort_order (sortOrder) to GraphQL schema

evansiroky committed:
Fix issues with car/vehicle rental routing in arriveBy mode

Also add more rental info to response

evansiroky committed:
Condense logging for VehicleRentalUpdater

evansiroky committed:
Make vehicle rental GBFS work for only free-floating systems

evansiroky committed:
Account for possibility of vehicle rentals that are walked for a bit

landonreed committed:
Merge branch 'micromobility' into 2019-05-23

evansiroky committed:
Implement banned micromobility travel and dropoff area configuration

hunterowens commented:
very cool @evansiroky! 

one quick question - does this factor in time to device (ie, it is a 5 minute walk to the nearest jump bike) and unlock time?

evansiroky commented:
@hunterowens, yes. OTP is multimodal and is sorta based off of the AStar graph search algorithm except there are States that are stored in the priority queue. States are differentiated by a number of things including the current mode of travel. Therefore, a graph search can traverse the same edge in a walking state or vehicle rental state. See this example of a trip plan with a bike rental:

<img width="1359" alt="Screen Shot 2019-06-17 at 5 42 41 PM" src="https://user-images.githubusercontent.com/3112493/59645556-5528cc80-9127-11e9-8e93-2b1daa34dac5.png">

And yest, there are built-in assumptions of time and costs to pickup and dropoff a rental vehicle. See here: https://github.com/ibi-group/OpenTripPlanner/blob/3561e0e0dbf43c0cd99dac101a0b67a7817a5709/src/main/java/org/opentripplanner/routing/core/RoutingRequest.java#L235-L244

evansiroky committed:
Merge branch 'micromobility' into 2019-05-23

evansiroky committed:
Merge commit '71a27326edb5fb0153fbf51cfd505210144a783f' into 2019-05-23

landonreed commented on src/client/js/otp/config.js:
```diff
@@ -342,4 +342,7 @@ otp.config.modes = {
     "CAR_HAIL,WALK,TRAM,RAIL,SUBWAY,FUNICULAR,GONDOLA"     : _tr('TNC &amp; Rail'),
     "CAR_RENT,WALK,TRANSIT"     : _tr('Car Rental &amp; Transit'),
     "CAR_RENT,WALK,TRAM,RAIL,SUBWAY,FUNICULAR,GONDOLA"     : _tr('Car Rental &amp; Rail'),
+    "MICROMOBILITY": _tr('Micromobility'),
```
The name "Micromobility" feels a bit inaccurate. I would expect that to encompass both bikes and scooters. Also, it just feels kind of like a trendy word that gets tossed around because someone decided people would pay money when they heard it uttered.

evansiroky replied:
What word do you recommend instead? The place that I'm coming from with this is from renowned analyst Horace Dediu who coined the word with this definition:

> ...what micromobility is is a motorized vehicle.  Earlier I said micromobility is 500 kg. or less but I would amend that definition to include (a) having a motor and (b) being used for utility transport. 

You can read more in [this blog post](https://micromobility.io/blog/2019/1/21/micromobility-an-introduction).

landonreed commented on src/client/js/otp/modules/planner/PlannerModule.js:
```diff
@@ -362,6 +362,10 @@ otp.modules.planner.PlannerModule =
             if(this.numItineraries) queryParams.numItineraries = this.numItineraries;
             if(this.minTransferTime) queryParams.minTransferTime = this.minTransferTime;
             if(this.showIntermediateStops) queryParams.showIntermediateStops = this.showIntermediateStops;
+            if(this.watts) queryParams.watts = this.watts;
```
`wattage` seems like a better name for this query param.

landonreed commented on src/client/js/otp/modules/planner/PlannerModule.js:
```diff
@@ -362,6 +362,10 @@ otp.modules.planner.PlannerModule =
             if(this.minTransferTime) queryParams.minTransferTime = this.minTransferTime;
             if(this.showIntermediateStops) queryParams.showIntermediateStops = this.showIntermediateStops;
+            if(this.watts) queryParams.watts = this.watts;
+            if(this.weight) queryParams.weight = this.weight;
```
`weight` feels like a very ambiguous param name. Maybe `passengerWeight` is better? Or `passMass` (jk).

evansiroky replied:
This is intended to model the combined weight of the passenger, vehicle and all other belongings the passenger carries on the vehicle. Maybe I can rename to `combinedVehiclePassengerAndBelongingsWeight`?

landonreed replied:
Nah, weight is fine.

landonreed commented on src/client/js/otp/modules/planner/PlannerModule.js:
```diff
@@ -362,6 +362,10 @@ otp.modules.planner.PlannerModule =
             if(this.showIntermediateStops) queryParams.showIntermediateStops = this.showIntermediateStops;
+            if(this.watts) queryParams.watts = this.watts;
+            if(this.weight) queryParams.weight = this.weight;
+            if(this.minimumMicromobilitySpeed) queryParams.minimumMicromobilitySpeed = this.minimumMicromobilitySpeed;
```
Is min speed really necessary?

evansiroky replied:
The reason I put minSpeed in here is to cover the case for when it would be possible to walk with the vehicle on grades so steep that the motor couldn't propel the vehicle up the hill. Like imagine if you had a kid's hoverboard with a weak motor... the speed calculations would say it'd only be able to go like 0.something miles per hour, but you could walk with the thing and still cover some ground. On the flipside, maybe you have a really heavy vehicle that you can't even push up a hill. In that case you'd want to set 0 as the minimum speed.

landonreed commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -397,6 +397,26 @@
     @QueryParam("onlyTransitTrips")
     private Boolean onlyTransitTrips;
 
+    // the amount of watts a Micromobility vehicle can sustainably output
```
Probably ought to follow above convention and use javadoc comments for these params. Ditto for below params.

evansiroky replied:
Will do

landonreed commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -397,6 +397,26 @@
+
+    // the minimum speed of a micromobility vehicle. This should only be used to avoid unreasonably slow times on hills.
+    // If it is desired to model effectively impossible travel uphill (ie the vehicle can't reasonably be transported
+    // up a steep enough grade, enter 0. Value in m/s.
```
This comment mixes up parens and commas. Should probably start all comments with capital letters. If there is no slope is the scooter speed assumed to be constant (based on some calculation with weight and wattage, I suppose)? I think that should be stated clearly.

evansiroky replied:
If there is no elevation data for a StreetEdge, the slope is assumed to be 0. I'll fix up the comments here.

landonreed commented on src/main/java/org/opentripplanner/api/parameter/QualifiedMode.java:
```diff
@@ -51,6 +51,12 @@ public void applyToRoutingRequest (RoutingRequest req, boolean usingTransit) {
         }
+        if (this.mode == TraverseMode.MICROMOBILITY) {
+            if (this.qualifiers.contains(Qualifier.RENT)) {
+                req.modes.setMode(TraverseMode.WALK, true); // turn on WALK for micromobility rental mode
```
Ah, I was wondering about this. I wonder if the mode selection in https://github.com/ibi-group/OpenTripPlanner/pull/2/files#diff-c792e5489240daf4ed1a16cc9eb98cde should have a comment indicating that this is the behavior on the backend (auto addition of walk mode for bike/micromobility rental).

evansiroky replied:
Sure, I'll add some comments there.

landonreed commented on src/main/java/org/opentripplanner/api/resource/VehicleRental.java:
```diff
@@ -0,0 +1,94 @@
+
+@Path("/routers/{routerId}/vehicle_rental")
+@XmlRootElement
+public class VehicleRental {
```
Needs a comment, especially if it is going to called the very generic `Vehicle`-rental.

evansiroky replied:
Will add

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/DefaultWayPropertySetSource.java:
```diff
@@ -369,6 +369,8 @@ public void populateProperties(WayPropertySet props) {
         props.setCarSpeed("highway=track", 4.5f); // ~= 10 mph
         props.setCarSpeed("highway=road", 11.2f); // ~= 25 mph
+        props.setCarSpeed("surface=gravel", 4.5f); // ~= 10 mph
+        props.setCarSpeed("surface=dirt", 4.5f); // ~= 10 mph
```
This change probably doesn't have relevance for micromobility stuff, right? I'm OK to leave it in, just wondering.

evansiroky replied:
It does because the Micromobility speed is capped at the maximum allowed car speed. I was seeing some shortcuts along lots of gravel roads being taken when trying out some routes, so I added this.

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OSMFilter.java:
```diff
@@ -99,9 +99,9 @@ public static StreetTraversalPermission getPermissionsForEntity(OSMWithTags enti
+            permission = permission.remove(StreetTraversalPermission.BICYCLE_AND_MICROMOBILITY);
         } else if (entity.isBicycleExplicitlyAllowed()) {
-            permission = permission.add(StreetTraversalPermission.BICYCLE);
+            permission = permission.add(StreetTraversalPermission.BICYCLE_AND_MICROMOBILITY);
```
I don't feel like this will lead to entirely accurate routing. For example, are there not cases where bicycles are explicitly allowed, but scooters are not? There may be other permissions that are inaccurate here as well.

evansiroky replied:
Right, there are certainly places where bikes are allowed, but eScooters are not. However, there are currently no OSM tags for this, so I assumed that micromobility travel is allowed where bikes are allowed. As part of the TriMet MOD project, PBOT sent us some geojson files of where eScootering is forbidden and that is loaded in over here https://github.com/ibi-group/OpenTripPlanner/pull/2/files#r296006280.

I'll add a comment to the code here that discusses this.

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java:
```diff
@@ -192,6 +207,37 @@ public void buildGraph(Graph graph, HashMap<Class<?>, Object> extra) {
         return (T) value;
     }
 
+    public void loadMicromobilityTravelRestrictions(GraphBuilderParameters params) {
```
I think this method's javadoc ought to perhaps have some info about the type of file expected for applying the restrictions.

evansiroky replied:
Will add

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java:
```diff
@@ -192,6 +207,37 @@ public void buildGraph(Graph graph, HashMap<Class<?>, Object> extra) {
+        }
+    }
+
+    private PreparedGeometry parseGeometry(String areaUrlOrFile) {
```
Could use javadoc about the input format and perhaps when it will return null.

evansiroky replied:
Will add

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java:
```diff
@@ -268,6 +314,10 @@ public void buildGraph(HashMap<Class<?>, Object> extra) {
+
+            // this extra iteration is needed to ensure that StreetEdges created from areas and other items are properly
+            // accounted for with micromobility restrictions
+            applyMircomobilityRestrictions(graph);
```
Typo in method name. Also, it isn't immediately clear that this is an extra iteration in the context of its calling. Might be worth mentioning where else this is called.

evansiroky replied:
Will correct spelling and point to other location where this method is called.

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java:
```diff
@@ -676,10 +726,39 @@ private void buildBasicGraph() {
             } // END loop over OSM ways
         }
 
+        // forbid travel with micromobility on edges that intersects with a restricted area
```
Please use javadoc format.

evansiroky replied:
Will do

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OpenStreetMapModule.java:
```diff
@@ -676,10 +726,39 @@ private void buildBasicGraph() {
+            }
+        }
+
+        // forbid travel with micromobility on edges that intersects with a restricted area
```
Please use javadoc format.

evansiroky replied:
Will do

landonreed commented on src/main/java/org/opentripplanner/routing/algorithm/AStar.java:
```diff
@@ -346,17 +346,20 @@ public void setTraverseVisitor(TraverseVisitor traverseVisitor) {
+        List<GraphPath> acceptedPaths = new LinkedList<>();
+        for (State targetAcceptedState : runState.targetAcceptedStates) {
+            if (targetAcceptedState.isFinal()) {
+                acceptedPaths.add(new GraphPath(targetAcceptedState, true));
```
I very much like these variable name refactors.

landonreed commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -130,11 +130,15 @@
     public boolean showIntermediateStops = false;
 
-    /** max walk/bike speed along streets, in meters per second */
+    /** max modal speed along streets, in meters per second */
```
Modal feels kind of jargony/confusing. Could you say `max speed per travel mode...` instead?

evansiroky replied:
I'll add comments to each of the variable names this comment refers to.

landonreed commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -211,12 +215,6 @@
-
-    /** Cost of getting on and off your own car */
-    public int carSwitchCost;
-
```
Are these unused or being replaced?

evansiroky replied:
They were unused.

landonreed commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -499,6 +519,12 @@
     public double minimumCarRentalDistance = 804.672;
 
+    /**
+     * A mimum travel distance to travel using a vehicle rental.
```
typo

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -527,6 +559,21 @@
+    /**
+     * the maximum sustained power output of a micromobility vehicle in watts
+     * Theoretically, this could also model human power.
+     * The default is set to 250 watts.
```
Might be worth giving an example of speed given watts and passenger weight.

evansiroky replied:
Well, speed depends on a whole host of other factors. I'll add some examples of micromobility vehicles and how many watts they have.

landonreed commented on src/main/java/org/opentripplanner/routing/core/State.java:
```diff
@@ -114,8 +123,8 @@ public State(Vertex vertex, Edge backEdge, long timeSeconds, long startTime, Rou
         // this should be harmless since reversed clones are only used when routing has finished
         this.stateData.opt = options;
         this.stateData.startTime = startTime;
+        this.time = timeSeconds * 1000;
```
Why was this assignment moved up? Is it important that it be here? If so, that probably deserves a comment.

evansiroky replied:
It was moved up because additional time can be added to the start time if beginning a trip in a TNC vehicle that has an ETA associated with it. 

I'll add a comment.

landonreed commented on src/main/java/org/opentripplanner/routing/core/StateData.java:
```diff
@@ -92,11 +99,19 @@
     // The ids of cars that have been rented so far
     protected Set<String> rentedCars = new HashSet<>();
 
+    // The ids of cars that have been rented so far
```
copypasta

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/RentAVehicleAbstractEdge.java:
```diff
@@ -0,0 +1,83 @@
+         * networks ("catch-all" network defined).
+         */
+        if (stationNetworks == null || rentedNetworks == null)
+            return true; // Always a match
```
If the conditional is bracketless, the return statement should be on the same line.

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/RentAVehicleAbstractEdge.java:
```diff
@@ -0,0 +1,83 @@
+        /*
+         * Two stations are compatible if they share at least one network. Special case for "null"
+         * networks ("catch-all" network defined).
+         */
```
I don't love the multi-line comment style within a method body.

evansiroky replied:
Will change

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/RentAVehicleOffEdge.java:
```diff
@@ -0,0 +1,72 @@
+        s1e.incrementWeight(options.vehicleRentalDropoffCost);
+        s1e.incrementTimeInSeconds(options.vehicleRentalDropoffTime);
+        State s1 = s1e.makeState();
+        return s1;
```
I feel like this method could use a couple of terse comments in the body. Maybe that's just my limited experience with OTP showing.

evansiroky replied:
I'll add some more comments. If you have "limited experience with OTP", then there definitely should be comments to help others who have even less experience. ;)

evansiroky committed:
Refactor Micromobility rental transitions on StreetEdges

- Differentiate between two possible scenarios of states that can be forked depending on the circumstance
1. eScooter encounters area with banned eScooter travel and allowed dropoff and can either continue renting the scooter, but walk with it or end the scooter rental.
2. Banned parking dropoff, but travel allowed, so traveler can either end the vehicle rental and walk, or continue rental in hopes of finding a legal place to park later.

Fixes https://github.com/ibi-group/trimet-mod-otp/issues/204

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -135,12 +134,19 @@
     // whether or not this street is a good place to dropoff a floating car rental
     private boolean suitableForFloatingCarRentalDropoff = true;
 
+    // whether or not this street is a good place to dropoff a floating car rental
```
copypasta

evansiroky replied:
will fix

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -403,31 +411,93 @@ else if (
+                        // make sure a vehicle hasn't already been rented after transit
+                        ? !s0.stateData.hasRentedCarPostTransit()
+                        // make sure a vehicle hasn't already been rented before transit
+                        : !s0.stateData.hasRentedCarPreTransit())
```
The comments for this ternary feel a little unclear in the context. Might be more helpful to have a single comment for the entire ternary: `If we have boarded transit, make sure car wasn't rented after transit trip. Otherwise, check that car was not rented already.`. The large comment for this conditional on line 408 could also probably use some tweaking.

evansiroky replied:
I actually want to move this stuff into the `isCarRentalDropoffAllowed`. I'll do that instead.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -403,31 +411,93 @@ else if (
+                    }
+                }
+            }
+            // possible transition to dropping off a floating vehicle when in "arrive by" mode
```
I guess this is a copy of the above car rental code. Is the plan to factor out car rental into the more generic vehicle rental, which would also apply to bikes and scooters?

evansiroky replied:
Yeah, the utlimate idea I have is to combine the car, vehicle and bike rental stuff over here. If you notice, there currently isn't any floating bike transition code on a StreetEdge because the bike rental code always assumes that a bike will be dropped off at a rental station and not on some random street.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     * http://www.kreuzotter.de/english/espeed.htm goes into great detail regarding this and presents the following
+     * equation which also appears on other websites such as https://www.gribble.org/cycling/power_v_speed.html.
+     *
+     * P = Cm * V * (Cd * A * ρ/2 * (V + W) ^ 2 + Frg + V * Crvn)
```
#math

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     * @param minSpeed The minimum speed that the micromobility should travel at in cases where the slope is too steep
+     *                 or the vehicle has ran out of energy.
+     * @param maxSpeed The maximum speed the vehicle can travel at on steep downhills.
+     * @return the travel time in seconds
```
The method name is `calculateMicromobilitySpeed ` but the return value is time in seconds.

evansiroky replied:
Oops, forgot to change this. In the first iteration of this code, it did in fact return the travel time.

Will fix.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     * Then, using the cardanic formulae, the following solutions are found:
+     *
+     * a = (W^3 - Crvn^3) / 27 - (W * (5 * W * Crvn + (8 * Crvn^2) / (Cd * A * ρ) - 6 * Frg)) / (9 * Cd * A * ρ) + (2 * Frg * Crvn) / (3 * (Cd * A * ρ)^2) + P / (Cm * Cd * A * ρ)
+     * b = (2 / (9 * Cd * A * ρ)) * (3 * Frg - 4 * W * Crvn - (W^2) * Cd * A * ρ/2 - (2 * Crvn) / (Cd * A * ρ))
```
What are `a` and `b`? They seem to appear out of nowhere.

evansiroky replied:
Unfortunately, all of the intermediate steps are not shown on [the website where I got this equation from](http://www.kreuzotter.de/english/espeed.htm#forml). And I can't find any other supporting evidence beyond that website for how or why this works. However, before coding this equation, I put the equation into a spreadsheet with varying power, weight and grades and it all seemed to make sense.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     *
+     * Although it could be possible to try to estimate wind speed using weather data, for now wind speed is assumed to
+     * be 0. Therefore, the above equations can be changed to exlude the parts where the wind speed being 0 would cause
+     * various components to no longer be needed. Thus these solutions are used:
```
I think we can very reasonably assume wind resistance is zero. Not sure if all of this work needs to be shown, but I suppose it doesn't hurt.

evansiroky replied:
I could see wind resistance being a huge factor in places like Holland, so I definitely think it's worth it to leave it in here in case someone wants to try to make an updater that assigns wind speed to edges in the future.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     * If a^2 + b^3 < 0:
+     * V = 2 * sqrt(-b) * cos((1 / 3) * arccos(a / sqrt(-b^3)) - (2 / 3) * (Crvn / (Cd * A * ρ))
+     *
+     * @param watts The sustained power ouptut in watts
```
Typo. Also, given that you reduce the watts in the method, it might be worth mentioning that this is the theoretical max (or whatever is a better way to describe it).

evansiroky replied:
This input of watts is an area that I think could be expanded upon quite a bit. If you think about it, on a steep uphill road an eScooter rider will probably be applying the maximum sustained power possible. However, on that road in reverse, the eScooter rider will most likely be applying the brakes, thus implying negative power.

I'll fix the typo and add this discussion to the comments.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     * @param Cr The coefficient of rolling resistance. This can also be used to model the difficulty of traveling over
+     *           bumpy roadways. See this wikipedia page for a list of coefficients by various surface types:
+     *           https://en.wikipedia.org/wiki/Rolling_resistance#Rolling_resistance_coefficient_examples
+     * @param Cdap The product of the coefficient of aerodynamic drag, frontal area and air density
```
Cr and Cdap violate conventions for naming variables (should begin with lowercase).

evansiroky replied:
I will spell out these variables and get rid of this abbreviated naming convention.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+     *
+     * V^3 + V^2 * 2 * (W + Crvn / (Cd * A * ρ)) + V * (W^2 + (2 * Frg) / (Cd * A * ρ)) - (2 * P) / (Cm * Cd * A * ρ) = 0
+     *
+     * Then, using the cardanic formulae, the following solutions are found:
```
I have no idea what the cardanic formulae is.

evansiroky replied:
It's a way to come up with solutions to cubic functions. Unfortunately, all of the intermediate steps are not shown on [the website where I got this equation from](http://www.kreuzotter.de/english/espeed.htm#forml).

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+        double Crvn = ElevationUtils.getDynamicRollingResistance(beta);
+
+        // Rolling friction (normalized on inclined plane) plus slope pulling force on inclined plane
+        double Frg = ElevationUtils.GRAVITATIONAL_ACCELERATION_CONSTANT *
```
Again, why do these variables begin with uppercase letters?

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -761,6 +854,142 @@ public double timeLowerBound(RoutingRequest options) {
+            minSpeed
+        );
+
+        return speed;
```
Yea, I still have no idea if this returns speed or time.

evansiroky replied:
It returns speed.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java:
```diff
@@ -1134,4 +1377,18 @@ public boolean containsCarNetwork(String carNetwork) {
+    }
+
+    public boolean containsVehicleNetwork(String vehicleNetwork) {
+        if (vehicleNetworks == null){
```
I thought there was some check above that said if any network set was null that the network was said to be contained. Maybe this is a separate check, but it might deserve a comment.

evansiroky replied:
Hmmm, I think you're right. However, I think moving forward it may be best to always require a vehicle network to be associated with a rental. Also, this may be in need of a mode-dependent rental network in a future refactor. I may just keep it the way it is for now.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetTraversalPermission.java:
```diff
@@ -69,7 +76,9 @@ public boolean allows(StreetTraversalPermission perm) {
-        } else if (modes.getBicycle() && allows(StreetTraversalPermission.BICYCLE)) {
+        } else if ((modes.getBicycle()) && allows(StreetTraversalPermission.BICYCLE)) {
+            return true;
+        } else if ((modes.getMicromobility()) && allows(StreetTraversalPermission.MICROMOBILITY)) {
```
Why the extra parens around getBicycle and getMicro?

evansiroky replied:
Will fix.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetVehicleRentalLink.java:
```diff
@@ -0,0 +1,102 @@
+
+    public State traverse(State s0) {
+        // Do not even consider vehicle rental vertices unless vehicle rental is enabled.
+        if ( ! s0.getOptions().allowVehicleRental) {
```
extra spaces?

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetWithElevationEdge.java:
```diff
@@ -90,6 +110,54 @@ public double getSlopeWorkCostEffectiveLength() {
+     * along a road are pre-calculated and allocated into bins of 1% grade. Two different arrays are analyzed. One array
+     * has information about the gradient and the other has information about the length in meters of this gradient. All
+     * of the resulting travel times and meters of each gradient segment are added up and then used to calculate an
+     * average speed for the entire road.
```
Are the elevation gradients used to modify bike speed as well?

Edit: looks like they're not, but it might be worth adding a TODO for this.

evansiroky replied:
Will add. Theoretically, this micromobility speed calculation could be used for bicycling as well.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetWithElevationEdge.java:
```diff
@@ -90,6 +110,54 @@ public double getSlopeWorkCostEffectiveLength() {
+    @Override
+    public double calculateSpeed(RoutingRequest options, TraverseMode traverseMode, long timeMillis) {
+        if (traverseMode != TraverseMode.MICROMOBILITY) return super.calculateSpeed(options, traverseMode, timeMillis);
+        // calculate the travel time it would tak to traverse each gradient
```
typo

evansiroky replied:
will fix and will also fix it to say it calculates speed and not travel time.

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetWithElevationEdge.java:
```diff
@@ -90,6 +110,54 @@ public double getSlopeWorkCostEffectiveLength() {
+        double time = 0;
+
+        // TODO: figure out why this is null sometimes
+        if (gradients == null) return super.calculateSpeed(options, traverseMode, timeMillis);
```
Might be good to move this short circuit higher up or combine with the traverseMode check.

evansiroky replied:
Will do

landonreed commented on src/main/java/org/opentripplanner/routing/edgetype/StreetWithElevationEdge.java:
```diff
@@ -90,6 +110,54 @@ public double getSlopeWorkCostEffectiveLength() {
+        return Math.min(
+            distance / time,
+            // make sure the speed limit is obeyed
+            getCarSpeed()
```
There are three instances where the speed is throttled by car speed. I think this should be refactored so that it only happens once (probably in a single return statement):

```
const edgeSpeed = time == 0 ? options.maximumMicromobilitySpeed : distance / time
return Math.min(edgeSpeed, getCarSpeed())
```
^^ add your comments in there

evansiroky replied:
The car speed is needed in the above loop of speed calculations per gradient bin to make sure that the speed used to calculate travel time for that gradient bin never exceeds the car speed. Now, given that it'll be impossible to ever achieve a travel time derived from a speed higher than the carSpeed, it should be possible to remove this last check down here.

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
 
+    // Coefficient for velocity-dependent dynamic rolling resistance, here approximated with 0.1
+    // See http://www.kreuzotter.de/english/espeed.htm
+    private static final double CrV = 0.1;
```
Usually static constants are `NAMED_LIKE_THIS`. I'm also not crazy about a variable name that can't be understood without prior knowledge. Why not call this `COEF_ROLLING_RESISTANCE` or something?

evansiroky replied:
Will rename.

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
+     * that much, so we just use the constant at 45 degrees of latitude.
+     * See https://en.wikipedia.org/wiki/Gravitational_acceleration#Gravity_model_for_Earth
+     */
+    public static final double GRAVITATIONAL_ACCELERATION_CONSTANT = 9.80665;
```
Phew, I haven't encountered this number since freshman year of undergrad.

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
+     * Fdrag = 0.5 * Cd * A * Rho * V^2
+     *               ⎣CdA_⎦
+     *
+     * See See https://www.gribble.org/cycling/power_v_speed.html
```
See See

evansiroky replied:
fix fix

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
+     * Cd = 0.63
+     * A = 0.6
+     */
+    private static final double CdA = 0.63 * 0.6;
```
Hard-to-understand variable name

evansiroky replied:
Will rename

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
+     * @param altitude The altitude in meters
+     */
+    public static double getDragResistiveForceComponent(double altitude) {
+        double randomlyGuessedTemperature = A_RANDOM_OUTDOOR_TEMPERATURE_IN_KELVIN - 9.8 * altitude / 1000;
```
I know it's explained above, but `9.8 / 1000` is a magic number here. Also, all things being equal, what is the difference in speed due to drag at say 1000m vs sea level? It can't be that great, right? This all feels a bit overkill to come up with a travel speed for a scooter.

evansiroky replied:
Yeah, the aerodynamic drag is not that great at lower speeds, but it makes a large difference the faster you go. However, the current OTP calculations for bicycling [do take elevation into account](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-1.x/src/main/java/org/opentripplanner/routing/util/ElevationUtils.java#L142), so I thought I should try to be as correct as possible here. At least I didn't try to calculate humidity also!

But I will remove the magic numbers.

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -21,6 +21,136 @@
+    private static final double SPECIFIC_GAS_CONSTANT_FOR_DRY_AIR = 287.058;
+
+    /**
+     * Calculates the componens of drag resistance except for the velocity assuming travel through dry earthy air. The
```
typo

evansiroky replied:
will fix

landonreed commented on src/main/java/org/opentripplanner/routing/util/ElevationUtils.java:
```diff
@@ -58,16 +188,57 @@ public static SlopeCosts getSlopeCosts(CoordinateSequence elev, boolean slopeLim
+                gradients.add(bin);
+            }
+
             // Baldwin St in Dunedin, NZ, is the steepest street
```
OK, now I want to visit Baldwin St: https://www.youtube.com/watch?v=zNCgYaYpH_4

landonreed commented on src/main/java/org/opentripplanner/routing/vehicle_rental/VehicleRentalStation.java:
```diff
@@ -0,0 +1,127 @@
+import java.util.Locale;
+import java.util.Set;
+
+public class VehicleRentalStation implements Serializable, Cloneable {
```
Probably deserves javadoc

evansiroky replied:
Will add

landonreed commented on src/main/java/org/opentripplanner/routing/vehicle_rental/VehicleRentalStationService.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Map;
+import java.util.Set;
+
+public class VehicleRentalStationService implements Serializable {
```
javadoc

evansiroky replied:
Will add

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+                }
+
+                // parse the region and update the regions
+                ObjectMapper mapper = new ObjectMapper();
```
Superfluous ObjectMapper instance? Can we just use the one that is a field on the class?

evansiroky replied:
Will do

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+public class GenericGbfsService implements VehicleRentalDataSource, JsonConfigurable {
+    private static final Logger LOG = LoggerFactory.getLogger(GenericGbfsService.class);
+
+    private ObjectMapper mapper = new ObjectMapper();
```
I think this can be static (only one needed by each service instance).

landonreed replied:
Not sure it really matters though.

evansiroky replied:
Will make static and final.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+     * the following is ideally a bit of temporary code that reads in the contents from a URL (or file) that describes
+     * the allowable dropoff areas for this vehicle rental company
+     *
+     * ideally in the future, the url will be built into the GBFS
```
Might be good to add a FIXME/TODO here if we expect a future change.

evansiroky replied:
Will add.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+        if (regionsUrlNode != null) {
+            regionsUrl = regionsUrlNode.asText();
+        }
+        JsonNode regionGeoJson = config.get("regionsGeoJson");
```
Variable name doesn't match field name from JSON.

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+    /**
+     * Construct a url based on the root url and the desired file
+     */
+    private String makeUrl(String file) {
```
There might be a more descriptive name for this method than `makeUrl`.

evansiroky replied:
Will rename to `makeGbfsEndpointUrl`.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+        String systemAlertsUrl = null;
+
+        if (rootData == null) {
+            // populate URLs with default values
```
Might be worth describing this condition a little better. Is `rootData == null` an allowed state per the GBFS spec? Or is this an attempt to fix the feed because it isn't constructed properly?

evansiroky replied:
Yeah, the root `gbfs.json` file/url is optional per [the GBFS spec](https://github.com/NABSA/gbfs/blob/master/gbfs.md#files).

Will add a comment about this.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+            systemPricingPlansUrl = makeUrl("system_pricing_plans.json");
+            systemAlertsUrl = makeUrl("system_alerts.json");
+        } else {
+            GbfsRespone gbfsRespone = null;
```
typo

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GBFSMappings/StationInformation.java:
```diff
@@ -0,0 +1,43 @@
+public class StationInformation extends BaseGtfsResponse {
+    public StationInfromation data;
+
+    public static class StationInfromation {
```
typo

evansiroky replied:
Will fix

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+    }
+
+    private void updateRegions() {
+        // TODO implement
```
Is this TODO here because it needs to be a part of this PR? Or is this dependent on GBFS changes? Or something else? Maybe just provide a little more info.

evansiroky replied:
Will add more comments

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+        try {
+            data = getDataFromUrlOrFile(url, headerName, headerValue);
+        } catch (IOException e) {
+            LOG.warn("Failed to fetch from url: {}. Error: {}", url, e);
```
Just curious -- how does `e` get rendered in the log? Is it legible? Does it need to use `Exception#getMessage` or some other?

evansiroky replied:
It'd be whatever e.toString() is implemented as. It could be verbose.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+        InputStream data = fetchFromUrl(url, true);
+        if (data == null) {
+            LOG.error("failed to fetch required data from system_information URL for feed: {}", networkName);
+            return;
```
I feel like maybe this needs to be refactored to use `fetchAndParseFromUrl`? There seems to be some repeated code.

evansiroky replied:
Yeah, there are some custom error messages depending on the type of failure.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/GenericGbfsService.java:
```diff
@@ -0,0 +1,536 @@
+        } else if (regionsUrl != null) {
+            try {
+                // fetch regions
+                InputStream data = fetchFromUrl(regionsUrl);
```
Should this be using fetchAndParseFromUrl? Edit: perhaps not, since it looks to have special parsing.

evansiroky replied:
Yeah, different parsing is used.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalDataSource.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+public interface VehicleRentalDataSource {
```
javadoc

evansiroky replied:
Will add.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalUpdater.java:
```diff
@@ -0,0 +1,339 @@
+
+import static org.opentripplanner.graph_builder.linking.SimpleStreetSplitter.DESTRUCTIVE_SPLIT;
+
+public class VehicleRentalUpdater extends PollingGraphUpdater {
```
javadoc

evansiroky replied:
Will add.

landonreed commented on src/main/java/org/opentripplanner/updater/vehicle_rental/VehicleRentalUpdater.java:
```diff
@@ -0,0 +1,339 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(VehicleRentalUpdater.class);
+
+    private static DecimalFormat format = new DecimalFormat("##.000000");
```
Might could make this final and give a better name. Something like `TO_SIX_DECIMALS` or `SUB_METER_PRECISION`.

evansiroky replied:
Will do.

landonreed reviewed (commented):
Sorry there are so many comments. Most of them are related to style or comments. I can't write too intelligently about a lot of the low level stuff happening with states and such. One other high-level comment is that some of the calculations for scooter speed (e.g., resistive drag force at elevation) seem a little overkill -- I'd like to be convinced that the difference in this force on a human traveling on a scooter has a significant impact on travel speed. If you can make that argument successfully, it probably also belongs embedded in code comments.

evansiroky committed:
Merge branch 'trimet-dev' into micromobility

evansiroky committed:
Add more docs, comments and fix some typos.

evansiroky committed:
Consolidate floating dropoff code that checks pre-post transit rentals

evansiroky committed:
Add comments and improve readability

evansiroky committed:
Fix weightLowerBound function to check for Micromobility mode

evansiroky committed:
Add comments and improve readability of micromobility speed calculations

evansiroky committed:
Update comments and improve readability for vehicle rental updater code

evansiroky commented:
Most of the comments have been addressed. However, I want to check that one change to the elevation calculations that I made are ok by rebuilding the graph. I'll do that either tomorrow night or first thing Monday.

evansiroky commented:
Ok, I think this is ready for review again.

landonreed reviewed (approved):
Nice work, @evansiroky!
